File: backend/src/controllers/attendanceController.js
import prisma from '../utils/db.js';

// Get local Cambodia (UTC+7) Date string and Time string
const getLocalTimeDetails = (customTime, customDate) => {
  const now = new Date();
  const cambodiaTime = new Date(now.toLocaleString("en-US", { timeZone: "Asia/Phnom_Penh" }));

  let dateString;
  if (customDate) {
    dateString = customDate; // Format: "YYYY-MM-DD"
  } else {
    const year = cambodiaTime.getFullYear();
    const month = String(cambodiaTime.getMonth() + 1).padStart(2, '0');
    const day = String(cambodiaTime.getDate()).padStart(2, '0');
    dateString = `${year}-${month}-${day}`;
  }

  let timeString;
  if (customTime) {
    timeString = customTime; // Format: "HH:mm"
  } else {
    const hours = String(cambodiaTime.getHours()).padStart(2, '0');
    const minutes = String(cambodiaTime.getMinutes()).padStart(2, '0');
    timeString = `${hours}:${minutes}`;
  }

  return {
    dateObj: new Date(dateString),
    dateString,
    timeString
  };
};

// Check-in or Check-out Logging
export const logCheckInOut = async (req, res) => {
  const { staffId, action, customTime, customDate, note } = req.body;

  if (!staffId || !action) {
    return res.status(400).json({ message: 'Staff ID and Action are required' });
  }

  const validActions = ['checkin_1', 'checkout_1', 'checkin_2', 'checkout_2'];
  if (!validActions.includes(action)) {
    return res.status(400).json({ message: 'Invalid action. Must be checkin_1, checkout_1, checkin_2, or checkout_2' });
  }

  try {
    // 1. Get Employee to verify shifts
    const employee = await prisma.employee.findUnique({
      where: { staffId }
    });

    if (!employee) {
      return res.status(404).json({ message: 'Employee not found' });
    }

    const { dateObj, timeString } = getLocalTimeDetails(customTime, customDate);

    // 2. Find or Create today's attendance record
    let attendance = await prisma.attendance.findUnique({
      where: {
        staffId_attendanceDate: {
          staffId,
          attendanceDate: dateObj
        }
      }
    });

    const updateData = {};
    if (note) updateData.note = note;

    if (action === 'checkin_1') {
      updateData.checkin1 = timeString;
    } else if (action === 'checkout_1') {
      updateData.checkout1 = timeString;
    } else if (action === 'checkin_2') {
      updateData.checkin2 = timeString;
    } else if (action === 'checkout_2') {
      updateData.checkout2 = timeString;
    }

    if (!attendance) {
      // Create new record
      attendance = await prisma.attendance.create({
        data: {
          staffId,
          attendanceDate: dateObj,
          ...updateData
        }
      });
    } else {
      // Update existing record
      attendance = await prisma.attendance.update({
        where: {
          id: attendance.id
        },
        data: updateData
      });
    }

    // 3. Recalculate isLate and isEarlyLeave
    const c1 = attendance.checkin1 || updateData.checkin1;
    const c2 = attendance.checkin2 || updateData.checkin2;
    const o1 = attendance.checkout1 || updateData.checkout1;
    const o2 = attendance.checkout2 || updateData.checkout2;

    const s1Start = employee.shift1Start;
    const s1End = employee.shift1End;
    const s2Start = employee.shift2Start;
    const s2End = employee.shift2End;

    let isLate = false;
    let isEarlyLeave = false;

    // Check-in 1 late check
    if (c1 && c1 > s1Start) {
      isLate = true;
    }
    // Check-in 2 late check
    if (c2 && c2 > s2Start) {
      isLate = true;
    }
    // Check-out 1 early check
    if (o1 && o1 < s1End) {
      isEarlyLeave = true;
    }
    // Check-out 2 early check
    if (o2 && o2 < s2End) {
      isEarlyLeave = true;
    }

    // Save recalculated metrics
    const finalAttendance = await prisma.attendance.update({
      where: { id: attendance.id },
      data: {
        isLate,
        isEarlyLeave
      },
      include: {
        employee: {
          select: {
            nameEn: true,
            nameKh: true,
            branch: true,
            department: { select: { nameEn: true, nameKh: true } }
          }
        }
      }
    });

    res.json({
      message: `Successfully logged ${action} at ${timeString}`,
      data: finalAttendance
    });
  } catch (error) {
    console.error('Log Check-in/out error:', error);
    res.status(500).json({ message: 'Server error logging attendance' });
  }
};

// Get today's attendance summary (for Admin/HR Dashboard)
export const getTodayAttendance = async (req, res) => {
  const { departmentId, branch } = req.query;
  const { dateObj } = getLocalTimeDetails();

  try {
    const where = {
      attendanceDate: dateObj
    };

    if (departmentId || branch) {
      where.employee = {};
      if (departmentId) where.employee.departmentId = departmentId;
      if (branch) where.employee.branch = branch;
    }

    const logs = await prisma.attendance.findMany({
      where,
      include: {
        employee: {
          select: {
            staffId: true,
            nameEn: true,
            nameKh: true,
            gender: true,
            branch: true,
            shift1Start: true,
            shift1End: true,
            shift2Start: true,
            shift2End: true,
            department: { select: { nameEn: true, nameKh: true } },
            position: { select: { titleEn: true, titleKh: true } }
          }
        }
      },
      orderBy: { createdAt: 'desc' }
    });

    res.json(logs);
  } catch (error) {
    console.error('Get today attendance error:', error);
    res.status(500).json({ message: 'Server error retrieving today attendance' });
  }
};

// Get attendance history (for reports & user logging history)
export const getHistory = async (req, res) => {
  const { staffId, startDate, endDate, departmentId, branch } = req.query;

  try {
    const where = {};

    if (staffId) {
      where.staffId = staffId;
    }

    if (departmentId || branch) {
      where.employee = where.employee || {};
      if (departmentId) where.employee.departmentId = departmentId;
      if (branch) where.employee.branch = branch;
    }

    if (startDate || endDate) {
      where.attendanceDate = {};
      if (startDate) {
        where.attendanceDate.gte = new Date(startDate);
      }
      if (endDate) {
        where.attendanceDate.lte = new Date(endDate);
      }
    }

    const history = await prisma.attendance.findMany({
      where,
      include: {
        employee: {
          select: {
            staffId: true,
            nameEn: true,
            nameKh: true,
            gender: true,
            branch: true,
            shift1Start: true,
            shift1End: true,
            shift2Start: true,
            shift2End: true,
            department: { select: { nameEn: true, nameKh: true } },
            position: { select: { titleEn: true, titleKh: true } }
          }
        }
      },
      orderBy: [
        { attendanceDate: 'desc' },
        { staffId: 'asc' }
      ]
    });

    res.json(history);
  } catch (error) {
    console.error('Get attendance history error:', error);
    res.status(500).json({ message: 'Server error retrieving attendance history' });
  }
};

// Get stats summary (Total present, late, early leave, on leave) for Dashboard
export const getStatsSummary = async (req, res) => {
  const { dateObj } = getLocalTimeDetails();

  try {
    // Total employees count
    const totalEmployees = await prisma.employee.count({
      where: { status: 'Active' }
    });

    // Attendance logs for today
    const todayLogs = await prisma.attendance.findMany({
      where: { attendanceDate: dateObj }
    });

    const presentCount = todayLogs.length;
    const lateCount = todayLogs.filter(log => log.isLate).length;
    const earlyLeaveCount = todayLogs.filter(log => log.isEarlyLeave).length;

    // Active approved leaves today
    const leavesToday = await prisma.leave.count({
      where: {
        leaveDate: dateObj,
        status: 'Approved'
      }
    });

    res.json({
      totalEmployees,
      presentToday: presentCount,
      lateToday: lateCount,
      earlyLeaveToday: earlyLeaveCount,
      onLeaveToday: leavesToday
    });
  } catch (error) {
    console.error('Get attendance stats error:', error);
    res.status(500).json({ message: 'Server error retrieving attendance stats' });
  }
};

// PUT /api/attendances/:id
export const updateAttendance = async (req, res) => {
  const { id } = req.params;
  const { checkin1, checkout1, checkin2, checkout2, note } = req.body;

  try {
    const attendance = await prisma.attendance.findUnique({
      where: { id },
      include: { employee: true }
    });

    if (!attendance) {
      return res.status(404).json({ message: 'Attendance record not found' });
    }

    const updated = await prisma.attendance.update({
      where: { id },
      data: {
        checkin1: checkin1 || null,
        checkout1: checkout1 || null,
        checkin2: checkin2 || null,
        checkout2: checkout2 || null,
        note: note || null
      }
    });

    const c1 = updated.checkin1;
    const c2 = updated.checkin2;
    const o1 = updated.checkout1;
    const o2 = updated.checkout2;

    const emp = attendance.employee;
    let isLate = false;
    let isEarlyLeave = false;

    if (c1 && c1 > emp.shift1Start) isLate = true;
    if (c2 && c2 > emp.shift2Start) isLate = true;
    if (o1 && o1 < emp.shift1End) isEarlyLeave = true;
    if (o2 && o2 < emp.shift2End) isEarlyLeave = true;

    const finalAttendance = await prisma.attendance.update({
      where: { id },
      data: { isLate, isEarlyLeave },
      include: {
        employee: {
          select: {
            staffId: true,
            nameEn: true,
            nameKh: true,
            department: { select: { nameEn: true, nameKh: true } }
          }
        }
      }
    });

    res.json({
      message: 'Attendance record updated successfully',
      data: finalAttendance
    });
  } catch (error) {
    console.error('Update attendance error:', error);
    res.status(500).json({ message: 'Server error updating attendance record' });
  }
};

// POST /api/attendances
export const createAttendance = async (req, res) => {
  const { staffId, attendanceDate, checkin1, checkout1, checkin2, checkout2, note } = req.body;

  if (!staffId || !attendanceDate) {
    return res.status(400).json({ message: 'Staff ID and Attendance Date are required' });
  }

  try {
    const employee = await prisma.employee.findUnique({
      where: { staffId }
    });

    if (!employee) {
      return res.status(404).json({ message: 'Employee not found' });
    }

    const dateObj = new Date(attendanceDate);

    // Check existing
    const existing = await prisma.attendance.findUnique({
      where: {
        staffId_attendanceDate: {
          staffId,
          attendanceDate: dateObj
        }
      }
    });

    if (existing) {
      return res.status(400).json({ message: 'Attendance record already exists for this date. Please edit instead.' });
    }

    const created = await prisma.attendance.create({
      data: {
        staffId,
        attendanceDate: dateObj,
        checkin1: checkin1 || null,
        checkout1: checkout1 || null,
        checkin2: checkin2 || null,
        checkout2: checkout2 || null,
        note: note || null
      }
    });

    const c1 = created.checkin1;
    const c2 = created.checkin2;
    const o1 = created.checkout1;
    const o2 = created.checkout2;

    let isLate = false;
    let isEarlyLeave = false;

    if (c1 && c1 > employee.shift1Start) isLate = true;
    if (c2 && c2 > employee.shift2Start) isLate = true;
    if (o1 && o1 < employee.shift1End) isEarlyLeave = true;
    if (o2 && o2 < employee.shift2End) isEarlyLeave = true;

    const finalAttendance = await prisma.attendance.update({
      where: { id: created.id },
      data: { isLate, isEarlyLeave },
      include: {
        employee: {
          select: {
            staffId: true,
            nameEn: true,
            nameKh: true,
            department: { select: { nameEn: true, nameKh: true } }
          }
        }
      }
    });

    res.status(201).json({
      message: 'Attendance record created successfully',
      data: finalAttendance
    });
  } catch (error) {
    console.error('Create attendance error:', error);
    res.status(500).json({ message: 'Server error creating attendance record' });
  }
};

// DELETE /api/attendances/:id
export const deleteAttendance = async (req, res) => {
  const { id } = req.params;

  try {
    const existing = await prisma.attendance.findUnique({
      where: { id }
    });

    if (!existing) {
      return res.status(404).json({ message: 'Attendance record not found' });
    }

    await prisma.attendance.delete({
      where: { id }
    });

    res.json({ message: 'Attendance record deleted successfully' });
  } catch (error) {
    console.error('Delete attendance error:', error);
    res.status(500).json({ message: 'Server error deleting attendance record' });
  }
};

File: backend1/src/main/java/com/hrchomnan/backend1/controller/AttendanceController.java
package com.hrchomnan.backend1.controller;

import com.hrchomnan.backend1.enums.LeaveStatus;
import com.hrchomnan.backend1.model.Attendance;
import com.hrchomnan.backend1.model.Employee;
import com.hrchomnan.backend1.repository.AttendanceRepository;
import com.hrchomnan.backend1.repository.EmployeeRepository;
import com.hrchomnan.backend1.repository.LeaveRepository;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@RestController
@RequestMapping("/attendances")
@RequiredArgsConstructor
public class AttendanceController {

    private final AttendanceRepository attendanceRepository;
    private final EmployeeRepository employeeRepository;
    private final LeaveRepository leaveRepository;

    @Data
    public static class LogRequest {
        private String staffId;
        private String action; // checkin_1, checkout_1, checkin_2, checkout_2
        private String customTime;
        private String customDate;
        private String note;
    }

    @PostMapping("/log")
    public ResponseEntity<?> logCheckInOut(@RequestBody LogRequest request) {
        if (request.getStaffId() == null || request.getAction() == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "Staff ID and Action are required"));
        }

        Optional<Employee> empOpt = employeeRepository.findByStaffId(request.getStaffId());
        if (empOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Employee not found"));
        }

        Employee employee = empOpt.get();
        LocalDate attendanceDate = request.getCustomDate() != null ? LocalDate.parse(request.getCustomDate()) : LocalDate.now();
        String timeString = request.getCustomTime() != null ? request.getCustomTime() : LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm"));

        Optional<Attendance> existingOpt = attendanceRepository.findByStaffIdAndAttendanceDate(employee.getStaffId(), attendanceDate);
        Attendance attendance = existingOpt.orElseGet(() -> Attendance.builder()
                .staffId(employee.getStaffId())
                .attendanceDate(attendanceDate)
                .build());

        if (request.getNote() != null) {
            attendance.setNote(request.getNote());
        }

        switch (request.getAction()) {
            case "checkin_1" -> attendance.setCheckin1(timeString);
            case "checkout_1" -> attendance.setCheckout1(timeString);
            case "checkin_2" -> attendance.setCheckin2(timeString);
            case "checkout_2" -> attendance.setCheckout2(timeString);
            default -> {
                return ResponseEntity.badRequest().body(Map.of("message", "Invalid action"));
            }
        }

        // Calculate late & early leave
        boolean isLate = false;
        boolean isEarlyLeave = false;

        if (attendance.getCheckin1() != null && attendance.getCheckin1().compareTo(employee.getShift1Start()) > 0) {
            isLate = true;
        }
        if (attendance.getCheckin2() != null && attendance.getCheckin2().compareTo(employee.getShift2Start()) > 0) {
            isLate = true;
        }
        if (attendance.getCheckout1() != null && attendance.getCheckout1().compareTo(employee.getShift1End()) < 0) {
            isEarlyLeave = true;
        }
        if (attendance.getCheckout2() != null && attendance.getCheckout2().compareTo(employee.getShift2End()) < 0) {
            isEarlyLeave = true;
        }

        attendance.setIsLate(isLate);
        attendance.setIsEarlyLeave(isEarlyLeave);

        Attendance saved = attendanceRepository.save(attendance);
        return ResponseEntity.ok(Map.of(
                "message", "Successfully logged " + request.getAction() + " at " + timeString,
                "data", saved
        ));
    }

    @GetMapping("/today")
    public ResponseEntity<List<Attendance>> getTodayAttendance() {
        return ResponseEntity.ok(attendanceRepository.findByAttendanceDate(LocalDate.now()));
    }

    @GetMapping("/history")
    public ResponseEntity<List<Attendance>> getHistory(
            @RequestParam(required = false) String staffId,
            @RequestParam(required = false) String startDate,
            @RequestParam(required = false) String endDate
    ) {
        if (staffId != null) {
            return ResponseEntity.ok(attendanceRepository.findByStaffId(staffId));
        }
        if (startDate != null && endDate != null) {
            return ResponseEntity.ok(attendanceRepository.findByAttendanceDateBetween(
                    LocalDate.parse(startDate), LocalDate.parse(endDate)
            ));
        }
        return ResponseEntity.ok(attendanceRepository.findAll());
    }

    @GetMapping("/stats-summary")
    public ResponseEntity<Map<String, Object>> getStatsSummary() {
        long totalEmployees = employeeRepository.count();
        List<Attendance> todayLogs = attendanceRepository.findByAttendanceDate(LocalDate.now());

        long presentCount = todayLogs.size();
        long lateCount = todayLogs.stream().filter(Attendance::getIsLate).count();
        long earlyLeaveCount = todayLogs.stream().filter(Attendance::getIsEarlyLeave).count();
        long onLeaveToday = leaveRepository.findByStatus(LeaveStatus.Approved).stream()
                .filter(l -> l.getLeaveDate().equals(LocalDate.now()))
                .count();

        Map<String, Object> stats = new HashMap<>();
        stats.put("totalEmployees", totalEmployees);
        stats.put("presentToday", presentCount);
        stats.put("lateToday", lateCount);
        stats.put("earlyLeaveToday", earlyLeaveCount);
        stats.put("onLeaveToday", onLeaveToday);

        return ResponseEntity.ok(stats);
    }

    @PostMapping
    public ResponseEntity<?> createAttendance(@RequestBody Attendance attendance) {
        if (attendance.getStaffId() == null || attendance.getAttendanceDate() == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "Staff ID and Attendance Date are required"));
        }
        Attendance saved = attendanceRepository.save(attendance);
        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of("message", "Attendance record created successfully", "data", saved));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateAttendance(@PathVariable UUID id, @RequestBody Attendance updated) {
        return attendanceRepository.findById(id)
                .map(existing -> {
                    existing.setCheckin1(updated.getCheckin1());
                    existing.setCheckout1(updated.getCheckout1());
                    existing.setCheckin2(updated.getCheckin2());
                    existing.setCheckout2(updated.getCheckout2());
                    existing.setNote(updated.getNote());
                    Attendance saved = attendanceRepository.save(existing);
                    return ResponseEntity.ok(Map.of("message", "Attendance record updated successfully", "data", saved));
                })
                .orElse(ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Attendance record not found")));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteAttendance(@PathVariable UUID id) {
        if (attendanceRepository.existsById(id)) {
            attendanceRepository.deleteById(id);
            return ResponseEntity.ok(Map.of("message", "Attendance record deleted successfully"));
        }
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Attendance record not found"));
    }
}
